adapters/gamma/gamma.go

```diff
@@ -0,0 +1,184 @@
+
+	request := *requestIn
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		err := &errortypes.BadInput{
+			Message: "No impressions in the bid request",
+		}
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	errors := make([]error, 0, 1)
+
+	var bidderExt adapters.ExtImpBidder
+	err := json.Unmarshal(request.Imp[0].Ext, &bidderExt)
+	if err != nil {
+		err = &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	var gammaExt openrtb_ext.ExtImpGamma
+	err = json.Unmarshal(bidderExt.Bidder, &gammaExt)
+	if err != nil {
+		err = &errortypes.BadInput{
+			Message: "ext.bidder.publisher not provided",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	if gammaExt.PartnerID == "" {
+		err = &errortypes.BadInput{
+			Message: "PartnerID is empty",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	if gammaExt.ZoneID == "" {
+		err = &errortypes.BadInput{
+			Message: "ZoneID is empty",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+	if gammaExt.WebID == "" {
+		err = &errortypes.BadInput{
+			Message: "WebID is empty",
+		}
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	request.AT = 1 //Defaulting to first price auction for all prebid requests
+
+	thisURI := a.URI
+	thisURI = thisURI + "?id=" + gammaExt.PartnerID
+	thisURI = thisURI + "&zid=" + gammaExt.ZoneID
+	thisURI = thisURI + "&wid=" + gammaExt.WebID
+	thisURI = thisURI + "&bidid=" + request.ID
+	thisURI = thisURI + "&hb=pbmobile"
+	if request.Device != nil {
+		thisURI = thisURI + "&device_ip=" + request.Device.IP
+		thisURI = thisURI + "&device_model=" + request.Device.Model
+		thisURI = thisURI + "&device_os=" + request.Device.OS
+		if request.Device.UA != "" {
+			thisURI = thisURI + "&device_ua=" + url.QueryEscape(request.Device.UA)
+		}
+		thisURI = thisURI + "&device_ifa=" + request.Device.IFA
+	}
+	if request.App != nil {
+		thisURI = thisURI + "&app_id=" + request.App.ID
+		thisURI = thisURI + "&app_bundle=" + request.App.Bundle
+		thisURI = thisURI + "&app_name=" + request.App.Name
+	}
+
+	headers := http.Header{}
+	headers.Add("Accept", "*/*")
+	headers.Add("x-openrtb-version", "2.5")
+	if gammaExt.PartnerID == "bad-request-id" {
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+	}
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	addHeaderIfNonEmpty(headers, "Connection", "keep-alive")
+	addHeaderIfNonEmpty(headers, "cache-control", "no-cache")
+	addHeaderIfNonEmpty(headers, "Accept-Encoding", "gzip, deflate")
+
+	return []*adapters.RequestData{{
```

Comment:
I don't see where you pass along any `Imp.ID`'s to your bidder, so that you can bid against the various imps. You do pass in the `request.ID` but that is the ID of the overall request, it does not tie directly into any of the imp IDs.

Reply:
Ok I have passed the Imp.ID into bid request. Please double check and let me know if it's ok now.

Reply:
You will only ever bid on the first impression in a request?

Reply:
Yes, our system only process one impression per request. So if request has multiformat, the first impression will be processed.

Reply:
You are aware that you are returning a slice of request data? You could create multiple requests, one per Imp, and then the system will call `MakeBids()` for each request sent to your bidder that produced a response, and gather all the bids together. This would allow you to bid on all the impressions even though your backend system can only handle one impression at a time.

Reply:
I would like to confirm you only want to bid on the first imp of the impression, or are you working on a solution to submit one `RequestData` per imp so you can bid on all imps?

Reply:
We are working on a solution to bid on all imps. It takes a little changes on our system.

Reply:
Did you understand that you could bid on all imps with just a change to the adapter, and no changes to your system? (unless you wanted to include bidding on all the imps on  set, rather than bidding on each bid in the request in isolation).